tests/unit/dense_conjugate.cpp

```diff
@@ -0,0 +1,194 @@
+
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <vector>
+#include <memory>
+#include <complex>
+
+#include <alp.hpp>
+#include "../utils/print_alp_containers.hpp"
+
+constexpr float tol = 1.e-10;
+
+template< typename T >
+T random_value();
+
+template<>
+float random_value< float >() {
+	return static_cast< float >( rand() ) / RAND_MAX;
+}
+
+template<>
+std::complex< float > random_value< std::complex< float > >() {
+	const float re = random_value< float >();
+	const float im = random_value< float >();
+	return std::complex< float >( re, im );
+}
+
+template< typename MatrixType >
+void init_matrix( MatrixType &M ) {
+	// Temporary until proper matrix building is implemented
+	typedef typename MatrixType::value_type value_type;
+	alp::internal::setInitialized( M, true );
+	const size_t height = alp::ncols( M );
+	const size_t width = alp::nrows( M );
+	for( size_t r = 0; r < height; ++r ) {
+		for( size_t c = 0; c < width; ++c ) {
+			const value_type val = random_value< value_type >();
+			if( r < c ) {
+				alp::internal::access( M, alp::internal::getStorageIndex( M, r, c ) ) = val;
+				if( r != c ) {
+					alp::internal::access( M, alp::internal::getStorageIndex( M, c, r ) ) = grb::utils::is_complex< value_type >::conjugate( val );
+				}
+			} else if ( r == c ) {
+				alp::internal::access( M, alp::internal::getStorageIndex( M, r, c ) ) = std::real( val );
+			}
+		}
+	}
+}
+
+template<
+	typename MatrixType1,
+	typename MatrixType2,
+	typename T = typename MatrixType1::value_type,
+	typename Ring
+>
+alp::RC check_if_same( const MatrixType1 &A, const MatrixType2 &B, const Ring &ring ) {
+
+	alp::RC rc = alp::SUCCESS;
+
+	alp::Matrix< T, alp::structures::Square > E( nrows( A ) );
+	rc = rc ? rc : set( E, alp::Scalar< T >( ring.template getZero< T >() ) );
+
+	rc = rc ? rc : alp::foldl( E, A, ring.getAdditiveOperator() );
+	rc = rc ? rc : alp::foldl( E, B, alp::operators::subtract< T >() );
+
+	float fnorm = ring.template getZero< float >();
+	rc = rc ? rc : alp::eWiseLambda(
+		[ &fnorm, &ring ]( const size_t i, const size_t j, T &val ) {
+			(void) i;
+			(void) j;
+			const float valsquare = static_cast< float >( std::real( val * grb::utils::is_complex< T >::conjugate( val ) ) );
```

Comment:
I would rather use `std::norm( val )` then  `val * conjugate( val )`, because the correct type will be automatically returned (i.e. `double` or `float`).